### PR TITLE
feat(symbol-collector): Add `symbol-collector` target

### PR DESCRIPTION
**Pull request by iker-barriocanal**
The goal of the `symbol-collector` target is to continue automating the release of [`sentry-java`](https://github.com/getsentry/sentry-java) (Android and Java). https://github.com/getsentry/craft/pull/258 automated publishing to Maven Central, and this PR automates uploading symbols to the Android Bucket.

### CODE REVIEW

File: Dockerfile
Comment (iker-barriocanal) [inline Dockerfile:39]: @bruno-garcia you may want to use another version of the Symbol Collector.

File: README.md
Comment (AbhiPrasad) [inline README.md:1033]:
```suggestion
targets:
  - name: maven
    gradleCliPath: ./gradlew
    mavenCliPath: scripts/mvnw.cmd
    mavenSettingsPath: scripts/settings.xml
    mavenRepoId: ossrh
    mavenRepoUrl: https://oss.sonatype.org/service/local/staging/deploy/maven2/
    android:
      distDirRegex: /^sentry-android-.*$/
      fileReplaceeRegex: /\d\.\d\.\d(-SNAPSHOT)?/
      fileReplacerStr: release.aar
```

File: README.md
Comment (AbhiPrasad) [inline README.md:1039]:
```suggestion
The `symbol-collector` client needs to be available in the path.
```

File: src/targets/symbolCollector.ts
Comment (AbhiPrasad) [inline src/targets/symbolCollector.ts:52]: I would rather do a check for each and the error message include what config option is missing - makes it easier to debug.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (AbhiPrasad) [inline src/targets/__tests__/symbolCollector.test.ts:35]:
```suggestion
      name: 'symbol-collector',
```

File: src/targets/__tests__/symbolCollector.test.ts
Comment (AbhiPrasad) [inline src/targets/__tests__/symbolCollector.test.ts:26]: Setting a default value on the arg will mean we don't need the ternary expression below.

```suggestion
  customConfig: Record<string, unknown> = { testKey: 'testVal' } 
```

File: src/targets/symbolCollector.ts
Comment (AbhiPrasad) [inline src/targets/symbolCollector.ts:33]: This seems like this should be private

Comment (AbhiPrasad) [reply src/targets/symbolCollector.ts:33]: Let's also then not make tests directly assert on this object as well, feels fragile to me.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (AbhiPrasad) [inline src/targets/__tests__/symbolCollector.test.ts:131]: Why inline snapshot over just asserting on the array?

File: src/utils/system.ts
Comment (AbhiPrasad) [inline src/utils/system.ts:276]: Who uses this?

File: src/utils/githubApi.ts
Comment (AbhiPrasad) [inline src/utils/githubApi.ts:76]: Why have these methods been added to this file?

Comment (iker-barriocanal) [reply src/utils/githubApi.ts]: Added them in the past and totally forgot to remove them lol.

File: src/utils/system.ts
Comment (iker-barriocanal) [reply src/utils/system.ts]: Same here, forgot to remove them.

File: Dockerfile
Comment (bruno-garcia) [inline Dockerfile:28]: Given the image has .NET already (assuming it's because of `nuget` target), I can have a build of SymbolCollector that's 'runtime-dependant' which should reduce its size to just a few KBs.

Also, can we bump to 5.0?

```suggestion
  dotnet-sdk-5.0 \
```

File: Dockerfile
Comment (bruno-garcia) [reply Dockerfile:39]: I think we need to go with 1.2.1 for now (as you found, there's an issue since the .NET 5 upgrade on symbol collector native CLI builds:

```suggestion
  && symbol_collector_url=$(curl -s https://api.github.com/repos/getsentry/symbol-collector/releases/tags/1.2.1 | \
```

File: README.md
Comment (bruno-garcia) [inline README.md:1046]:
```suggestion
| `batchType`      | The batch type of the symbols to be uploaded. I.e: `Android`, `macOS`, `iOS`.                                                |
```

**Review by bruno-garcia — COMMENTED** [review]
sweet! Left a couple suggestions. Thanks @iker-barriocanal

File: Dockerfile
Comment (BYK) [inline Dockerfile:20]: Let's bring back the indent.

File: Dockerfile
Comment (BYK) [inline Dockerfile:40]: I don't know how much overhead `jq` brings but we can replace this with some `awk` and `grep` magic if we ever need to 😀

File: Dockerfile
Comment (BYK) [inline Dockerfile:42]: I think you can just pipe this into `unzip` and avoid storing on disk?

File: src/targets/symbolCollector.ts
Comment (BYK) [reply src/targets/symbolCollector.ts:52]: So much this.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:56]: Should this really be configurable? Also, we should probably not call this an endpoint but just a URL.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:63]: ??
```suggestion
    const bundleId = this.symbolCollectorConfig.bundleIdPrefix + version;
```

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:64]: You don't need this, `this.getArtifactsForRevision` already produces debug logs.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:70]: This is the second place we write the exact same code. I think this code should be moved into `getArtifactsForRevision` itself.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:72]:
```suggestion
    if (artifacts.length === 0) {
```

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:73]: Shouldn't this be a warning?
```suggestion
      this.logger.warn(`Didn't found any artifacts after filtering`);
```

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:87]: Don't be gimmicky please :)
```suggestion
          const subdirPath = join(dir, String(index));
```

Also, instead of using the index, can we not use `artifact.name` or something?

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:107]: `spawnProcess` should be doing this already.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (BYK) [inline src/targets/__tests__/symbolCollector.test.ts:110]: Are you sure these `lastUpdated` strings are valid? I think they should be ISO-8601 date strings?

File: src/targets/__tests__/symbolCollector.test.ts
Comment (BYK) [reply src/targets/__tests__/symbolCollector.test.ts:131]: Snapshots are much easier to manage 🙂

**Review by BYK — APPROVED** [review]
Didn't look too deep into tests but overall LGTM, great job!

File: src/targets/symbolCollector.ts
Comment (AbhiPrasad) [reply src/targets/symbolCollector.ts:72]: We should turn on https://eslint.org/docs/rules/eqeqeq in craft

File: README.md
Comment (iker-barriocanal) [reply README.md:1046]: Added this, but it'd be best to have that documentation somewhere else and point there. Updating Craft's docs because the symbol collector got e.g. a new feature and now supports a new type doesn't seem ideal.

File: Dockerfile
Comment (iker-barriocanal) [reply Dockerfile:42]: `unzip` has a bug:

> Archives read from standard input are not yet supported, except with `funzip` (and then only the first member of the archive can be extracted).

`funzip` has some limitations:

> and it extracts the first member from the archive to stdout

(note the _first member_ and _to stdout_.)

I haven't found any alternatives _without_ storing on disk and directly piping, so suggestions are welcomed.

File: src/targets/symbolCollector.ts
Comment (iker-barriocanal) [reply src/targets/symbolCollector.ts:56]: Yes, this should be configurable. The reason why this is called endpoint (and not just URL or any other way), is because the option in the symbol collector is `server-endpoint`. I think it'd be misleading to not be consistent here.

File: src/targets/symbolCollector.ts
Comment (iker-barriocanal) [reply src/targets/symbolCollector.ts:87]: Do you have any suggestions to generate these unique names in a cheap way, without using the artifact names (and indexes), and before downloading the files?

File: src/targets/symbolCollector.ts
Comment (iker-barriocanal) [reply src/targets/symbolCollector.ts:33]: All the configs in the targets are public, and that's the reason why this is too. If we're going to make them private, I think we should make them all private and not have some that are public and some others that aren't.

> Let's also then not make tests directly assert on this object as well, feels fragile to me.

Any suggestions to test whether the target has the required config parameters? By looking at the codebase, I don't think having getters is better; if we're changing the conventions, we shouldn't do it for a single target. Wdyt?

Comment (AbhiPrasad) [reply src/targets/symbolCollector.ts:33]: Ok we should match the conventions - disregard my thoughts then. Maybe something to keep in mind if we decide to refactor the targets in the future.

In terms of the tests, I always prefer checking if the behaviour enabled by different config options is working rather than asserting on the config option itself. Asserting on resulting behaviour produces stronger tests and also matches real life usage much better - don’t find much value from just expecting the config property to be a certain value.

File: Dockerfile
Comment (iker-barriocanal) [reply Dockerfile:28]: Opened another PR for that: https://github.com/getsentry/craft/pull/267

File: src/artifact_providers/base.ts
Comment (AbhiPrasad) [inline src/artifact_providers/base.ts:91]: Do we need the undefined check if the properties (ex. `includesNames`) for `ParsedFilterOptions` are typed as optional?

File: src/artifact_providers/base.ts
Comment (AbhiPrasad) [inline src/artifact_providers/base.ts:67]: Please add a note that the raw filter options should be processed by `parseFilterOptions`.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (AbhiPrasad) [inline src/targets/__tests__/symbolCollector.test.ts:26]: We don't need to use an explicit string key.

```suggestion
  config: Record<string, unknown> = { testKey: 'testVal' }
```

File: src/targets/symbolCollector.ts
Comment (iker-barriocanal) [reply src/targets/symbolCollector.ts:32]: The behavior is either throwing an error (because the config wasn't correct) or successfully running the `publish` step. Not throwing an error is covered (there are tests with incorrect config expecting it to throw errors), and the `publish` step has its own tests too. Is there something that I'm missing?

**Review by AbhiPrasad — APPROVED** [review]
Nice! I would wait for Burak's answer about `unzip` before merging, but otherwise looks solid

File: Dockerfile
Comment (BYK) [reply Dockerfile:42]: Oh, I wasn't aware of that. I guess I was thinking of `tar` as that's the more common format across Linux :)

File: src/artifact_providers/base.ts
Comment (BYK) [inline src/artifact_providers/base.ts:78]: ❤️

File: src/artifact_providers/base.ts
Comment (BYK) [inline src/artifact_providers/base.ts:88]: ❤️ ❤️

File: src/artifact_providers/__tests__/base.test.ts
Comment (BYK) [inline src/artifact_providers/__tests__/base.test.ts:12]: I think we should also have one undefined one string combination.

Moreover, you can probably use `test.each` to clean up some of these cases. This is optional but the additional test cases is something we should definitely add.

File: src/targets/symbolCollector.ts
Comment (BYK) [reply src/targets/symbolCollector.ts:32]: As much as I agree with @AbhiPrasad in a general sense, I think for this specific use case, this is fine. As long as the config is readonly, it is good enough for now.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:39]: This seems like a common pattern so maybe we should abstract config parsing into the base class and make sure it is always called by the default constructor `parseTargetSpecificConfig()`

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:44]: This is not really checking a config unless you make the bin name configurable through an env variable or the config file itself.

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:70]: Why not pass `excludeNames` too?

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:78]: Should this logic be part of the base target class?

File: src/targets/symbolCollector.ts
Comment (BYK) [inline src/targets/symbolCollector.ts:90]: Might be better to just return the final promise:

```js
return this.artifactProvider.downloadArtifact(artifact, subdirPath);
```

File: src/targets/__tests__/symbolCollector.test.ts
Comment (BYK) [inline src/targets/__tests__/symbolCollector.test.ts:8]: We may wanna lift the common `fs` mocks into a global mock module as this pattern seems common enough now.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (BYK) [inline src/targets/__tests__/symbolCollector.test.ts:42]: Would have been better to make the executable path configurable and make sure we actually throw on a missing file.

File: src/targets/__tests__/symbolCollector.test.ts
Comment (BYK) [inline src/targets/__tests__/symbolCollector.test.ts:66]: I'd lift these checks into their own case as this doesn't seem relevant to the case we are testing.

**Pull request merged**
